chips/nrf52840/src/ieee802154_radio.rs

```diff
@@ -1146,23 +1347,47 @@ impl<'a> kernel::hil::radio::RadioData<'a> for Radio<'a> {
         buf: &'static mut [u8],
         frame_len: usize,
     ) -> Result<(), (ErrorCode, &'static mut [u8])> {
-        if self.tx_buf.is_some() || self.transmitting.get() {
+        if self.tx_buf.is_some() {
+            // tx_buf TakeCell is only occupied when a transmission is underway. This
+            // check insures we do not interrupt an ungoing transmission
             return Err((ErrorCode::BUSY, buf));
         } else if radio::PSDU_OFFSET + frame_len >= buf.len() {
             // Not enough room for CRC
             return Err((ErrorCode::SIZE, buf));
         }
 
+        if let RadioState::OFF = self.state.get() {
+            self.radio_initialize();
+        }
```

Comment:
Do we also need to `radio_initialize()` in the `start()` or `reset()` functions? If not, why do we need to do so here?

Reply:
I added `radio_initialize()` to the `start()` and `reset()` methods

Reply:
Is it possible for transmit() to work correctly if start() or reset() has not been called? If not, why do we still need to initialize here?

Reply:
I assumed that you could call `stop()` to turn off the radio, and then `start()` to start receiving again. Or `stop()` and then later `transmit()` to go straight to transmitting without hanging out in receive mode for a bit.

(I'm actually not sure why `reset()` exists at all, but it's part of the HIL, so maybe it makes more sense on other chips)

Reply:
Do others have thoughts on if transmit should fail if the radio is turned off? A case can be made for both implementations, and I am unsure which makes more sense.

Reply:
I think `transmit()` should turn on the radio AND successfully transmit. It shows intent from the client to use the radio, so just do the right thing. 

What Hudson was asking above was _whether_ this is the case or not. I.e., does `transmit()` work in the current radio state is OFF?

Reply:
> I think transmit() should turn on the radio AND successfully transmit.

@brghena I think this is misreading at least the current HIL (which unfortunately is underdocumented so we have to look at how it's used instead).

There are separate interfaces for managing the power state of the radio and for transmitting frames and `start` (power on) is asynchronous (the client waits for a state change notification in the `PowerClient` trait)---most likely because at least one implementation (the rf233) is over i2c, so any operation needs to be async. The _current_ MAC implementation (xmac) explicitly uses this---when it wants to transmit, if first calls `start` and then actually transmits from the power-on callback.

I think it's _fine_ for this implementation to do one better and turn the radio on for transmit, but probably _better_ to avoid it and push that functionality to the client---this would avoid buggy clients that only work on some hardware that happens to turn the radio on in `transmit`.

Reply:
Thanks for the higher-level thinking Amit! Looking closer, you're definitely right here about the intent of the HIL as designed.

That being said, the Radio HIL is only loosely implemented by this chip driver as it stands (in part because of how undocumented it is, I would guess). The `PowerClient` for example, isn't implemented and `start()` doesn't have a way to provide a callback. There is a question as to whether calling start and then transmit quickly would mess anything up, but I suspect not for this hardware implementation.

So what would be the best course of action for this PR? I don't think meeting every aspect of the HIL should be the immediate requirement since this PR is really just implementing ACKs.

Reply:
@brghena got it yes. `start` is still doing a fair amount of work and it's still that case that, but your point is well taken.

I'd suggest, in this PR, _not_ adding the above code that checks if the radio state is off and, if so, initializes the radio. Instead, subsequent changes can maybe either bring this chip implementation closer to something reasonable with the HIL (probably simply through a deferred call) or else change the HIL.

Reply:
@brghena @alevy Is there consensus on removing this?

Reply:
Consensus: leave this alone for now entirely. It's fine as-is.

Long-term: the 15.4 driver here should better implement the HIL which expects callbacks for power state changes. But that doesn't have to happen here. (And simultaneously we would be reviewing whether that HIL is appropriate or not)